test/Kestrel.Core.Tests/Http2ConnectionTests.cs

```diff
@@ -2886,6 +2954,158 @@ public async Task ConnectionResetNotLoggedWithNoActiveStreams()
+
+            await SendRstStreamAsync(1);
+
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+            VerifyGoAway(await ReceiveFrameAsync(), 1, Http2ErrorCode.NO_ERROR);
+
+            await WaitForAllStreamsAsync();
+        }
+
+        [Fact]
+        public async Task AcceptNewStreamsDuringClosingConnection()
+        {
+            await InitializeConnectionAsync(_echoApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: false);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
+
+            await SendGoAwayAsync();
+
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closing].Wait(5000));
+
+            await StartStreamAsync(3, _browserRequestHeaders, endStream: false);
+
+            Assert.Equal(Http2Connection.ConnectionState.Closing, _connection.State);
+            Assert.False(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+            Assert.Equal(3, _connection.HighestOpenedStreamId);
+
+            await SendDataAsync(1, _helloBytes, true);
+            await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 37,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 5,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 0,
+                withFlags: (byte)Http2DataFrameFlags.END_STREAM,
+                withStreamId: 1);
+            await SendDataAsync(3, _helloBytes, true);
+            await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 37,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 3);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 5,
+                withFlags: (byte)Http2DataFrameFlags.NONE,
+                withStreamId: 3);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 0,
+                withFlags: (byte)Http2DataFrameFlags.END_STREAM,
+                withStreamId: 3);
+
+            await WaitForConnectionStopAsync(expectedLastStreamId: 3, ignoreNonGoAwayFrames: false);
+        }
+
+        [Fact]
+        public async Task RejectNewStreamsDuringClosedConnection()
+        {
+            await InitializeConnectionAsync(_echoApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: false);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
```

Comment:
Is this racy?

Reply:
Yea and it's unnecessary since it will be checked via goaway frames sent later on.